bmp388/bmp388.go

```diff
@@ -0,0 +1,246 @@
+	}
+}
+
+// Configure can enable settings on the BMP388 and reads the calibration coefficients
+func (d *Device) Configure(config BMP388Config) (err error) {
+	d.Config = config
+
+	if d.Config == (BMP388Config{}) {
+		d.Config.Mode = NORMAL
+	}
+
+	if d.Config.SeaLevelPressure == 0 {
+		d.Config.SeaLevelPressure = 10132500 // standard sea level pressure in cPa
+	}
+
+	// Turning on the pressure and temperature sensors and setting the measurement mode
+	err = d.writeRegister(REG_PWR_CTRL, PWR_PRESS|PWR_TEMP|byte(d.Config.Mode))
+
+	// Configure the oversampling, output data rate, and iir filter coefficient settings
+	err = d.writeRegister(REG_OSR, byte(d.Config.Pressure|d.Config.Temperature<<3))
+	err = d.writeRegister(REG_ODR, byte(d.Config.ODR))
+	err = d.writeRegister(REG_IIR, byte(d.Config.IIR<<1))
+
+	if err != nil {
+		return errors.New("bmp388: failed to configure sensor, check connection")
+	}
+
+	// Check if there is a problem with the given configuration
+	if d.configurationError() {
+		return errors.New("bmp388: there is a problem with the configuration, try reducing ODR")
+	}
+
+	// Reading the builtin calibration coefficients and parsing them per the datasheet. The compensation formula given
+	// in the datasheet is implemented in floating point
+	buffer, err := d.readRegister(REG_CALI, 21)
+	if err != nil {
+		return errors.New("bmp388: failed to read calibration coefficient register")
+	}
+
+	d.cali.t1 = uint16(buffer[1])<<8 | uint16(buffer[0])
+	d.cali.t2 = uint16(buffer[3])<<8 | uint16(buffer[2])
+	d.cali.t3 = int8(buffer[4])
+
+	d.cali.p1 = int16(buffer[6])<<8 | int16(buffer[5])
+	d.cali.p2 = int16(buffer[8])<<8 | int16(buffer[7])
+	d.cali.p3 = int8(buffer[9])
+	d.cali.p4 = int8(buffer[10])
+	d.cali.p5 = uint16(buffer[12])<<8 | uint16(buffer[11])
+	d.cali.p6 = uint16(buffer[14])<<8 | uint16(buffer[13])
+	d.cali.p7 = int8(buffer[15])
+	d.cali.p8 = int8(buffer[16])
+	d.cali.p9 = int16(buffer[18])<<8 | int16(buffer[17])
+	d.cali.p10 = int8(buffer[19])
+	d.cali.p11 = int8(buffer[20])
+
+	return nil
+}
+
+// ReadTemperature returns the temperature in centicelsius, i.e 2426 / 100 = 24.26 C
+func (d *Device) ReadTemperature() (int32, error) {
+
+	rawTemp, err := d.readSensorData(REG_TEMP)
+	if err != nil {
+		return 0, err
+	}
+
+	// pulled from C driver: https://github.com/BoschSensortec/BMP3-Sensor-API/blob/master/bmp3.c
+	partialData1 := rawTemp - (256 * int64(d.cali.t1))
+	partialData2 := int64(d.cali.t2) * partialData1
+	partialData3 := (partialData1 * partialData1)
+	partialData4 := partialData3 * int64(d.cali.t3)
+	partialData5 := (partialData2 * 262144) + partialData4
+	partialData6 := partialData5 / 4294967296
+
+	d.cali.tlin = partialData6
+	temp := ((partialData6 * 25) / 16384)
+	return int32(temp), nil
+}
+
+// ReadPressure returns the pressure in centipascals, i.e 10132520 / 100 = 101325.20 Pa
+func (d *Device) ReadPressure() (int32, error) {
+
+	_, err := d.ReadTemperature()
+	if err != nil {
+		return 0, err
+	}
```

Comment:
Why are you reading the temperature here? Is reading the temperature required by the datasheet?

Reply:
Yes, there is a value called `tlin` from the temperature compensation calculations that the pressure calculations use too. I used `ReadTemperature` here because it already does all the compensation work to calculate `tlin` then returns the temperature by dividing by a constant. I'll refactor this to be less confusing.